packages/visual-editor/src/vite-plugin/plugin.ts

```diff
@@ -4,28 +4,137 @@ import { Plugin } from "vite";
 import mainTemplate from "./templates/main.tsx?raw";
 import editTemplate from "./templates/edit.tsx?raw";
 import directoryTemplate from "./templates/directory.tsx?raw";
+import { ComponentField, ComponentFieldMappings } from "@yext/visual-editor";
+import { defaultLayoutData } from "./defaultLayoutData.ts";
+
+type TemplateDefinition = {
+  name: string;
+  description: string;
+  content: string;
+  layoutRequired: boolean;
+  includeInManifest: boolean;
+  exampleSiteUrl?: string;
+  defaultLayoutData?: string;
+  componentFields?: Array<ComponentField>;
+};
 
 // files to generate
-const virtualFiles: Record<string, string> = {
-  "src/templates/main.tsx": mainTemplate,
-  "src/templates/edit.tsx": editTemplate,
-  "src/templates/directory.tsx": directoryTemplate,
+const templateDefinitions: Record<string, TemplateDefinition> = {
+  "src/templates/main.tsx": {
+    name: "main",
+    description:
+      "Use this template to generate pages for each of your Locations.",
+    content: mainTemplate,
+    layoutRequired: true,
+    includeInManifest: true,
+    exampleSiteUrl: "",
+    defaultLayoutData: defaultLayoutData.main,
+    componentFields: [
```

Comment:
Your example has `componentFieldMappings`. It would be better to model the template-manifest.json itself and you can just call JSON.stringify on the object you build up. You do that already sort of, but you have a secondary object on line 75 because you need to change this name and deal with the `content` part.

type VirtualFile {
  filepath: string;
  content: string;
  templateManifestEntry: TemplateManifestEntry;
}
...

// write template-manifest.json
const manifest = { templates: []}; (with typing of course)
virtualFiles.foreach(v => manifest.templates.push(v))
JSON.stringify(manifest)